modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java

```diff
@@ -126,16 +132,20 @@ public void layout(TextRun run, PGFont font, FontStrike strike, char[] text) {
             OSPango.pango_attr_list_insert(attrList, attr);
         }
 
-        if (str == 0L) {
-            str = OSPango.g_utf16_to_utf8(text);
+        Long str = runUtf8.get(run);
+        if (str == null) {
+            char[] rtext = Arrays.copyOfRange(text, run.getStart(), run.getEnd());
+            str = OSPango.g_utf16_to_utf8(rtext);
+            runUtf8.put(run, str);
             if (check(str, "Failed allocating UTF-8 buffer.", context, desc, attrList)) {
                 return;
             }
         }
 
         /* Itemize */
         long start = OSPango.g_utf8_offset_to_pointer(str, run.getStart());
-        long end = OSPango.g_utf8_offset_to_pointer(str, run.getEnd());
+        long utflen = OSPango.g_utf8_strlen(str,-1);
+        long end = OSPango.g_utf8_offset_to_pointer(str, utflen);
```

Comment:
Since you are now creating `n` native strings, 1 per substring based on the `TextRun`, rather than 1 for the entire String, isn't the `start` pointer wrong? Unless I am missing something, I would think that `start` should be set to `str`.

Reply:
I did a quick test, and setting `start = str` fixes the spurious assertions and intermittent crash.

Reply:
I should add that this is without any attempt to filter out `0` chars. Both `UnicodeTextTest` and  `UnicodeTextTest2` run correctly with no crashes and no assertions when I comment out the (ineffective) loop checking for 0 and set `start = str` leaving everything else as you have it in the current PR. Loading https://gluonhq.com/ in WebView works, too.

Reply:
I agree. the text array is now the subset for the run, so effectively run.getStart() should be zero,
so there should be no need for that call. This probably looks OK when there is one run not otherwise.

Reply:
You're both correct. I simplified it.

Reply:
Looks good now.